File: api/internal/schema/record_test.go
package schema

import (
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

type vectorFile struct {
	Format string `json:"format"`
	Version int   `json:"version"`
	Size    int   `json:"size"`
	CRC     struct {
		CheckInput string `json:"check_input"`
		Check      string `json:"check"`
	} `json:"crc"`
	Valid []struct {
		Name   string `json:"name"`
		Hex    string `json:"hex"`
		Record struct {
			Version uint8   `json:"version"`
			Type    uint8   `json:"type"`
			Quality uint8   `json:"quality"`
			Flags   uint8   `json:"flags"`
			Seq     uint32  `json:"seq"`
			Time    uint32  `json:"time"`
			Value   float32 `json:"value"`
			BootID  uint16  `json:"boot_id"`
		} `json:"record"`
	} `json:"valid"`
	Invalid []struct {
		Name  string `json:"name"`
		Hex   string `json:"hex"`
		Error string `json:"error"`
	} `json:"invalid"`
}

func loadVectors(t *testing.T) vectorFile {
	t.Helper()
	path := filepath.Join("..", "..", "..", "schema", "vectors", "records.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read vectors: %v", err)
	}
	var v vectorFile
	if err := json.Unmarshal(raw, &v); err != nil {
		t.Fatalf("parse vectors: %v", err)
	}
	if v.Format != "record" {
		t.Fatalf("unexpected vector format %q", v.Format)
	}
	return v
}

func mustHex(t *testing.T, s string) []byte {
	t.Helper()
	b, err := hex.DecodeString(s)
	if err != nil {
		t.Fatalf("bad hex %q: %v", s, err)
	}
	return b
}

func TestCRCCheckValue(t *testing.T) {
	v := loadVectors(t)
	got := CRC16CCITTFalse([]byte(v.CRC.CheckInput))
	want := mustHex(t, v.CRC.Check)
	if got != uint16(want[0])<<8|uint16(want[1]) {
		t.Fatalf("crc(%q) = %04x, want %s", v.CRC.CheckInput, got, v.CRC.Check)
	}
}

func TestConstantsMatchVectors(t *testing.T) {
	v := loadVectors(t)
	if v.Version != RecordVersion || v.Size != RecordSize {
		t.Fatalf("vectors are version %d size %d, code is version %d size %d",
			v.Version, v.Size, RecordVersion, RecordSize)
	}
}

func TestDecodeValidVectors(t *testing.T) {
	for _, vec := range loadVectors(t).Valid {
		t.Run(vec.Name, func(t *testing.T) {
			got, err := Decode(mustHex(t, vec.Hex))
			if err != nil {
				t.Fatalf("decode: %v", err)
			}
			want := Record{
				Version: vec.Record.Version,
				Type:    vec.Record.Type,
				Quality: vec.Record.Quality,
				Flags:   vec.Record.Flags,
				Seq:     vec.Record.Seq,
				Time:    vec.Record.Time,
				Value:   vec.Record.Value,
				BootID:  vec.Record.BootID,
			}
			if got != want {
				t.Fatalf("decode = %+v, want %+v", got, want)
			}
		})
	}
}

func TestEncodeValidVectorsByteForByte(t *testing.T) {
	for _, vec := range loadVectors(t).Valid {
		t.Run(vec.Name, func(t *testing.T) {
			r := Record{
				Version: vec.Record.Version,
				Type:    vec.Record.Type,
				Quality: vec.Record.Quality,
				Flags:   vec.Record.Flags,
				Seq:     vec.Record.Seq,
				Time:    vec.Record.Time,
				Value:   vec.Record.Value,
				BootID:  vec.Record.BootID,
			}
			got := Encode(r)
			if hex.EncodeToString(got[:]) != vec.Hex {
				t.Fatalf("encode = %x, want %s", got, vec.Hex)
			}
		})
	}
}

func TestInvalidVectorsReportNamedError(t *testing.T) {
	for _, vec := range loadVectors(t).Invalid {
		t.Run(vec.Name, func(t *testing.T) {
			_, err := Decode(mustHex(t, vec.Hex))
			if err == nil {
				t.Fatal("decode succeeded, want error")
			}
			if err.Error() != vec.Error {
				t.Fatalf("error = %q, want %q", err.Error(), vec.Error)
			}
		})
	}
}

func TestNames(t *testing.T) {
	if TypeName(1) != "soil_resistance_ohms" || TypeName(250) != "unknown" {
		t.Fatal("TypeName mismatch")
	}
	if QualityName(3) != "open" || QualityName(9) != "unknown" {
		t.Fatal("QualityName mismatch")
	}
}

File: app/src/protocol/__tests__/record.test.ts
import fs from 'fs';
import path from 'path';

import { fromBase64, fromHex, toBase64, toHex } from '../bytes';
import { crc16 } from '../crc16';
import { decodeRecord, encodeRecord, qualityName, readingTypeName, RECORD_SIZE, RECORD_VERSION, RecordDecodeError } from '../record';

type Vectors = {
  format: string;
  version: number;
  size: number;
  crc: { check_input: string; check: string };
  valid: { name: string; hex: string; record: { version: number; type: number; quality: number; flags: number; seq: number; time: number; value: number; boot_id: number } }[];
  invalid: { name: string; hex: string; error: string }[];
};

const vectors: Vectors = JSON.parse(
  fs.readFileSync(path.join(__dirname, '..', '..', '..', '..', 'schema', 'vectors', 'records.json'), 'utf8'),
);

describe('record codec against schema/vectors/records.json', () => {
  test('vector file is the record format we implement', () => {
    expect(vectors.format).toBe('record');
    expect(vectors.version).toBe(RECORD_VERSION);
    expect(vectors.size).toBe(RECORD_SIZE);
  });

  test('crc check value', () => {
    const input = new TextEncoder().encode(vectors.crc.check_input);
    expect(crc16(input).toString(16).padStart(4, '0')).toBe(vectors.crc.check);
  });

  test.each(vectors.valid.map((v) => [v.name, v] as const))('decodes %s', (_name, v) => {
    const r = decodeRecord(fromHex(v.hex));
    expect(r).toEqual({
      version: v.record.version,
      type: v.record.type,
      quality: v.record.quality,
      flags: v.record.flags,
      seq: v.record.seq,
      time: v.record.time,
      value: v.record.value,
      bootId: v.record.boot_id,
    });
  });

  test.each(vectors.valid.map((v) => [v.name, v] as const))('encodes %s byte for byte', (_name, v) => {
    const encoded = encodeRecord({
      version: v.record.version,
      type: v.record.type,
      quality: v.record.quality,
      flags: v.record.flags,
      seq: v.record.seq,
      time: v.record.time,
      value: v.record.value,
      bootId: v.record.boot_id,
    });
    expect(toHex(encoded)).toBe(v.hex);
  });

  test.each(vectors.invalid.map((v) => [v.name, v] as const))('rejects %s', (_name, v) => {
    expect(() => decodeRecord(fromHex(v.hex))).toThrow(RecordDecodeError);
    try {
      decodeRecord(fromHex(v.hex));
    } catch (e) {
      expect((e as RecordDecodeError).code).toBe(v.error);
    }
  });

  test('names', () => {
    expect(readingTypeName(1)).toBe('soil_resistance_ohms');
    expect(readingTypeName(250)).toBe('unknown');
    expect(qualityName(3)).toBe('open');
    expect(qualityName(9)).toBe('unknown');
  });
});

describe('base64 helpers', () => {
  test('round trips every vector and matches the standard alphabet', () => {
    for (const v of vectors.valid) {
      const bytes = fromHex(v.hex);
      const b64 = toBase64(bytes);
      expect(b64).toBe(Buffer.from(bytes).toString('base64'));
      expect(toHex(fromBase64(b64))).toBe(v.hex);
    }
  });

  test('handles lengths that need one or two padding characters', () => {
    for (const len of [0, 1, 2, 3, 4, 5, 19, 20, 21]) {
      const bytes = new Uint8Array(len).map((_, i) => (i * 37) & 0xff);
      expect(toBase64(bytes)).toBe(Buffer.from(bytes).toString('base64'));
      expect(Array.from(fromBase64(toBase64(bytes)))).toEqual(Array.from(bytes));
    }
  });
});
